File: workers/ruby/lib/tasker_core/step_handler/functional.rb
# frozen_string_literal: true

# Functional/block DSL for step handlers (TAS-294).
#
# This module provides block-based alternatives to the class-based handler API.
# It reduces boilerplate for common handler patterns while preserving full access
# to the underlying StepContext for advanced use cases.
#
# The DSL methods auto-wrap return values and classify exceptions:
# - Hash return -> StepHandlerCallResult.success(result: hash)
# - StepHandlerCallResult return -> pass through unchanged
# - PermanentError raised -> StepHandlerCallResult.error(retryable: false)
# - RetryableError raised -> StepHandlerCallResult.error(retryable: true)
# - Other errors -> StepHandlerCallResult.error(retryable: true)
#
# @example Basic handler
#   include TaskerCore::StepHandler::Functional
#
#   step_handler "process_payment",
#     depends_on: { cart: "validate_cart" },
#     inputs: [:payment_info] do |cart:, payment_info:, context:|
#
#     raise TaskerCore::Errors::PermanentError, "Payment info required" unless payment_info
#     result = charge_card(payment_info, cart["total"])
#     { payment_id: result["id"], amount: cart["total"] }
#   end

module TaskerCore
  module StepHandler
    module Functional
      # ========================================================================
      # Helper: Decision
      # ========================================================================

      # Helper for decision handler return values.
      #
      # @example
      #   decision_handler "route_order",
      #     depends_on: { order: "validate_order" } do |order:, context:|
      #     if order["tier"] == "premium"
      #       Decision.route(["process_premium"], tier: "premium")
      #     else
      #       Decision.route(["process_standard"])
      #     end
      #   end
      class Decision
        attr_reader :type, :steps, :reason, :routing_context

        def initialize(type, steps, reason, routing_context)
          @type = type
          @steps = steps
          @reason = reason
          @routing_context = routing_context
          freeze
        end

        # Route to the specified steps.
        #
        # @param steps [Array<String>] Step names to route to
        # @param routing_context [Hash] Additional routing context
        # @return [Decision]
        def self.route(steps, **routing_context)
          new('create_steps', Array(steps), nil, routing_context)
        end

        # Skip all branches.
        #
        # @param reason [String] Reason for skipping
        # @param routing_context [Hash] Additional routing context
        # @return [Decision]
        def self.skip(reason, **routing_context)
          new('no_branches', [], reason, routing_context)
        end
      end

      # ========================================================================
      # Helper: BatchConfig
      # ========================================================================

      # Configuration returned by batch analyzer handlers.
      #
      # @example
      #   batch_analyzer "analyze_orders", worker_template: "process_batch" do |context:|
      #     BatchConfig.new(total_items: 250, batch_size: 100)
      #   end
      class BatchConfig
        attr_reader :total_items, :batch_size, :metadata

        # @param total_items [Integer] Total number of items to process
        # @param batch_size [Integer] Size of each batch
        # @param metadata [Hash] Optional metadata
        def initialize(total_items:, batch_size:, metadata: {})
          @total_items = total_items
          @batch_size = batch_size
          @metadata = metadata
          freeze
        end
      end

      # ========================================================================
      # Internal: Auto-wrapping
      # ========================================================================

      module_function

      # @api private
      def _wrap_result(result)
        case result
        when Types::StepHandlerCallResult::Success,
             Types::StepHandlerCallResult::Error,
             Types::StepHandlerCallResult::CheckpointYield
          result
        when Hash
          Types::StepHandlerCallResult.success(result: result)
        when nil
          Types::StepHandlerCallResult.success(result: {})
        else
          Types::StepHandlerCallResult.success(result: result.respond_to?(:to_h) ? result.to_h : {})
        end
      end

      # @api private
      def _wrap_exception(exception)
        Types::StepHandlerCallResult.from_exception(exception)
      end

      # @api private
      def _inject_args(context, depends_on, inputs)
        args = { context: context }

        depends_on.each do |param_name, step_name|
          args[param_name.to_sym] = context.get_dependency_result(step_name.to_s)
        end

        inputs.each do |input_key|
          args[input_key.to_sym] = context.get_input(input_key.to_s)
        end

        args
      end

      # ========================================================================
      # Public DSL Methods
      # ========================================================================

      # Define a step handler from a block.
      #
      # Returns a StepHandler::Base subclass that can be registered with HandlerRegistry.
      # The block receives injected dependencies, inputs, and context as keyword arguments.
      # Return values are auto-wrapped as success results, and exceptions are
      # auto-classified as failure results.
      #
      # @param name [String] Handler name (must match step definition)
      # @param depends_on [Hash] Mapping of parameter names to dependency step names
      # @param inputs [Array<Symbol, String>] Task context input keys to inject
      # @param version [String] Handler version (default: "1.0.0")
      # @yield [**args] Block receiving dependencies, inputs, and context as keyword args
      # @return [Class] StepHandler::Base subclass
      #
      # @example
      #   ProcessPayment = step_handler "process_payment",
      #     depends_on: { cart: "validate_cart" },
      #     inputs: [:payment_info] do |cart:, payment_info:, context:|
      #
      #     raise TaskerCore::Errors::PermanentError, "Payment info required" unless payment_info
      #     result = charge_card(payment_info, cart["total"])
      #     { payment_id: result["id"], amount: cart["total"] }
      #   end
      def step_handler(name, depends_on: {}, inputs: [], version: '1.0.0', &block)
        raise ArgumentError, 'block required' unless block

        handler_depends = depends_on
        handler_inputs = inputs
        handler_block = block

        Class.new(Base) do
          const_set(:VERSION, version)

          define_method(:handler_name) { name }

          define_method(:call) do |context|
            args = Functional._inject_args(context, handler_depends, handler_inputs)
            result = handler_block.call(**args)
            Functional._wrap_result(result)
          rescue StandardError => e
            Functional._wrap_exception(e)
          end
        end
      end

      # Define a decision handler from a block.
      #
      # The block should return a `Decision.route(...)` or `Decision.skip(...)`.
      #
      # @param name [String] Handler name
      # @param depends_on [Hash] Mapping of parameter names to dependency step names
      # @param inputs [Array<Symbol, String>] Task context input keys to inject
      # @param version [String] Handler version (default: "1.0.0")
      # @yield [**args] Block returning a Decision
      # @return [Class] StepHandler::Base subclass with decision capabilities
      #
      # @example
      #   RouteOrder = decision_handler "route_order",
      #     depends_on: { order: "validate_order" } do |order:, context:|
      #     if order["tier"] == "premium"
      #       Decision.route(["process_premium"], tier: "premium")
      #     else
      #       Decision.route(["process_standard"])
      #     end
      #   end
      def decision_handler(name, depends_on: {}, inputs: [], version: '1.0.0', &block)
        raise ArgumentError, 'block required' unless block

        handler_depends = depends_on
        handler_inputs = inputs
        handler_block = block

        Class.new(Base) do
          include Mixins::Decision

          const_set(:VERSION, version)

          define_method(:handler_name) { name }

          define_method(:call) do |context|
            args = Functional._inject_args(context, handler_depends, handler_inputs)
            raw_result = handler_block.call(**args)

            case raw_result
            when Types::StepHandlerCallResult::Success,
                 Types::StepHandlerCallResult::Error
              raw_result
            when Decision
              result_data = {}
              if raw_result.type == 'create_steps'
                result_data[:routing_context] = raw_result.routing_context unless raw_result.routing_context.empty?
                decision_success(steps: raw_result.steps, result_data: result_data)
              else
                result_data[:reason] = raw_result.reason if raw_result.reason
                result_data[:routing_context] = raw_result.routing_context unless raw_result.routing_context.empty?
                decision_no_branches(result_data: result_data)
              end
            else
              Functional._wrap_result(raw_result)
            end
          rescue StandardError => e
            Functional._wrap_exception(e)
          end
        end
      end

      # Define a batch analyzer handler.
      #
      # The block should return a `BatchConfig` with `total_items` and `batch_size`.
      # Cursor configs are generated automatically.
      #
      # @param name [String] Handler name
      # @param worker_template [String] Name of the worker template step
      # @param depends_on [Hash] Mapping of parameter names to dependency step names
      # @param inputs [Array<Symbol, String>] Task context input keys to inject
      # @param version [String] Handler version (default: "1.0.0")
      # @yield [**args] Block returning a BatchConfig
      # @return [Class] StepHandler::Base subclass
      #
      # @example
      #   AnalyzeOrders = batch_analyzer "analyze_orders",
      #     worker_template: "process_batch" do |context:|
      #     BatchConfig.new(total_items: 250, batch_size: 100)
      #   end
      def batch_analyzer(name, worker_template:, depends_on: {}, inputs: [], version: '1.0.0', &block)
        raise ArgumentError, 'block required' unless block

        handler_depends = depends_on
        handler_inputs = inputs
        handler_block = block
        handler_worker_template = worker_template

        Class.new(Base) do
          include Mixins::Batchable

          const_set(:VERSION, version)

          define_method(:handler_name) { name }

          define_method(:call) do |context|
            args = Functional._inject_args(context, handler_depends, handler_inputs)
            raw_result = handler_block.call(**args)

            case raw_result
            when Types::StepHandlerCallResult::Success,
                 Types::StepHandlerCallResult::Error
              raw_result
            when BatchConfig
              total_items = raw_result.total_items
              batch_size = raw_result.batch_size
              worker_count = (total_items.to_f / batch_size).ceil

              cursor_configs = create_cursor_configs(total_items, worker_count)

              create_batches_outcome(
                worker_template_name: handler_worker_template,
                cursor_configs: cursor_configs,
                total_items: total_items,
                metadata: raw_result.metadata
              )
            else
              Functional._wrap_result(raw_result)
            end
          rescue StandardError => e
            Functional._wrap_exception(e)
          end
        end
      end

      # Define a batch worker handler.
      #
      # The block receives a `batch_context` keyword argument extracted from the
      # step context, containing cursor configuration for this worker's partition.
      #
      # @param name [String] Handler name
      # @param depends_on [Hash] Mapping of parameter names to dependency step names
      # @param inputs [Array<Symbol, String>] Task context input keys to inject
      # @param version [String] Handler version (default: "1.0.0")
      # @yield [**args] Block receiving batch_context and other injected args
      # @return [Class] StepHandler::Base subclass
      #
      # @example
      #   ProcessBatch = batch_worker "process_batch" do |batch_context:, context:|
      #     cursor = batch_context&.dig(:cursor_config)
      #     # process items from cursor[:start_cursor] to cursor[:end_cursor]
      #     { processed: true }
      #   end
      def batch_worker(name, depends_on: {}, inputs: [], version: '1.0.0', &block)
        raise ArgumentError, 'block required' unless block

        handler_depends = depends_on
        handler_inputs = inputs
        handler_block = block

        Class.new(Base) do
          include Mixins::Batchable

          const_set(:VERSION, version)

          define_method(:handler_name) { name }

          define_method(:call) do |context|
            args = Functional._inject_args(context, handler_depends, handler_inputs)

            # Delegate batch context extraction to Batchable mixin
            args[:batch_context] = get_batch_context(context)

            result = handler_block.call(**args)
            Functional._wrap_result(result)
          rescue StandardError => e
            Functional._wrap_exception(e)
          end
        end
      end
    end
  end
end

File: workers/typescript/src/handler/functional.ts
/**
 * Functional/factory API for step handlers (TAS-294).
 *
 * This module provides factory-function alternatives to the class-based handler API.
 * It reduces boilerplate for common handler patterns while preserving full access
 * to the underlying StepContext for advanced use cases.
 *
 * Factory functions auto-wrap return values and classify exceptions:
 * - Record return → StepHandlerResult.success(record)
 * - StepHandlerResult return → pass through unchanged
 * - PermanentError thrown → StepHandlerResult.failure(retryable=false)
 * - RetryableError thrown → StepHandlerResult.failure(retryable=true)
 * - Other errors → StepHandlerResult.failure(retryable=true)
 *
 * @example
 * ```typescript
 * import { defineHandler, PermanentError } from 'tasker-core';
 *
 * const ProcessPayment = defineHandler('process_payment', {
 *   depends: { cart: 'validate_cart' },
 *   inputs: { paymentInfo: 'payment_info' },
 * }, async ({ cart, paymentInfo, context }) => {
 *   if (!paymentInfo) throw new PermanentError('Payment info required');
 *   const result = await chargeCard(paymentInfo, cart.total);
 *   return { paymentId: result.id, amount: cart.total };
 * });
 * ```
 *
 * @module handler/functional
 */

import type { BatchWorkerContext } from '../types/batch.js';
import { createBatches as createBatchesOutcome } from '../types/batch.js';
import { ErrorType } from '../types/error-type.js';
import type { StepContext } from '../types/step-context.js';
import { StepHandlerResult } from '../types/step-handler-result.js';
import { StepHandler } from './base.js';
import { BatchableMixin, type RustCursorConfig } from './batchable.js';
import { type APICapable, applyAPI } from './mixins/api.js';
import { DecisionMixin } from './mixins/decision.js';

// ============================================================================
// Error Classes
// ============================================================================

/**
 * Base error for permanent, non-retryable failures.
 *
 * Throw this in functional handlers when the error will not succeed on retry.
 *
 * @example
 * ```typescript
 * throw new PermanentError('Invalid account number');
 * ```
 */
export class PermanentError extends Error {
  readonly retryable = false;
  readonly metadata: Record<string, unknown>;

  constructor(message: string, metadata?: Record<string, unknown>) {
    super(message);
    this.name = 'PermanentError';
    this.metadata = metadata ?? {};
  }
}

/**
 * Base error for transient, retryable failures.
 *
 * Throw this in functional handlers when a retry might succeed.
 *
 * @example
 * ```typescript
 * throw new RetryableError('Service temporarily unavailable');
 * ```
 */
export class RetryableError extends Error {
  readonly retryable = true;
  readonly metadata: Record<string, unknown>;

  constructor(message: string, metadata?: Record<string, unknown>) {
    super(message);
    this.name = 'RetryableError';
    this.metadata = metadata ?? {};
  }
}

// ============================================================================
// Helper Types
// ============================================================================

/**
 * Options for defineHandler and related factory functions.
 */
export interface HandlerOptions {
  /** Mapping of parameter names to dependency step names */
  depends?: Record<string, string>;
  /** Mapping of parameter names to input data keys */
  inputs?: Record<string, string>;
  /** Handler version (default: "1.0.0") */
  version?: string;
}

/**
 * The injected arguments object passed to functional handlers.
 *
 * Contains declared dependencies, inputs, and always includes `context`.
 */
export type HandlerArgs = Record<string, unknown> & {
  context: StepContext;
};

/**
 * Handler function signature.
 */
export type HandlerFn = (
  args: HandlerArgs
) => Promise<Record<string, unknown> | StepHandlerResult | undefined>;

/**
 * Helper for decision handler return values.
 *
 * @example
 * ```typescript
 * const RouteOrder = defineDecisionHandler('route_order', {
 *   depends: { order: 'validate_order' },
 * }, async ({ order }) => {
 *   if (order.tier === 'premium') {
 *     return Decision.route(['process_premium'], { tier: 'premium' });
 *   }
 *   return Decision.route(['process_standard']);
 * });
 * ```
 */
export class Decision {
  private constructor(
    readonly type: 'create_steps' | 'no_branches',
    readonly steps: string[],
    readonly reason?: string,
    readonly routingContext: Record<string, unknown> = {}
  ) {}

  /**
   * Route to the specified steps.
   */
  static route(steps: string[], routingContext?: Record<string, unknown>): Decision {
    return new Decision('create_steps', steps, undefined, routingContext ?? {});
  }

  /**
   * Skip all branches.
   */
  static skip(reason: string, routingContext?: Record<string, unknown>): Decision {
    return new Decision('no_branches', [], reason, routingContext ?? {});
  }
}

/**
 * Configuration returned by batch analyzer handlers.
 */
export interface BatchConfig {
  totalItems: number;
  batchSize: number;
  metadata?: Record<string, unknown>;
}

// ============================================================================
// Internal: Auto-wrapping
// ============================================================================

function wrapResult(result: unknown): StepHandlerResult {
  if (result instanceof StepHandlerResult) {
    return result;
  }
  if (result != null && typeof result === 'object' && !Array.isArray(result)) {
    return StepHandlerResult.success(result as Record<string, unknown>);
  }
  if (result == null) {
    return StepHandlerResult.success({});
  }
  return StepHandlerResult.success({ result });
}

function wrapError(error: unknown): StepHandlerResult {
  if (error instanceof PermanentError) {
    return StepHandlerResult.failure(
      error.message,
      ErrorType.PERMANENT_ERROR,
      false,
      error.metadata
    );
  }
  if (error instanceof RetryableError) {
    return StepHandlerResult.failure(
      error.message,
      ErrorType.RETRYABLE_ERROR,
      true,
      error.metadata
    );
  }
  const message = error instanceof Error ? error.message : String(error);
  return StepHandlerResult.failure(message, ErrorType.HANDLER_ERROR, true, {
    errorType: error instanceof Error ? error.constructor.name : typeof error,
  });
}

function injectArgs(
  context: StepContext,
  depends: Record<string, string>,
  inputs: Record<string, string>
): HandlerArgs {
  const args: HandlerArgs = { context };

  for (const [paramName, stepName] of Object.entries(depends)) {
    args[paramName] = context.getDependencyResult(stepName);
  }

  for (const [paramName, inputKey] of Object.entries(inputs)) {
    args[paramName] = context.getInput(inputKey);
  }

  return args;
}

// ============================================================================
// Public Factory Functions
// ============================================================================

/**
 * Define a step handler from a function.
 *
 * Returns a StepHandler subclass that can be registered with HandlerRegistry.
 * The function receives injected dependencies, inputs, and context.
 * Return values are auto-wrapped as success results, and exceptions are
 * auto-classified as failure results.
 *
 * @param name - Handler name (must match step definition)
 * @param options - Dependencies, inputs, and version
 * @param fn - Handler function
 * @returns StepHandler subclass
 *
 * @example
 * ```typescript
 * const ProcessPayment = defineHandler('process_payment', {
 *   depends: { cart: 'validate_cart' },
 *   inputs: { paymentInfo: 'payment_info' },
 * }, async ({ cart, paymentInfo, context }) => {
 *   if (!paymentInfo) throw new PermanentError('Payment info required');
 *   const result = await chargeCard(paymentInfo, cart.total);
 *   return { paymentId: result.id, amount: cart.total };
 * });
 *
 * // Register:
 * registry.register(ProcessPayment);
 * ```
 */
export function defineHandler(
  name: string,
  options: HandlerOptions,
  fn: HandlerFn
): typeof StepHandler & { new (): StepHandler } {
  const depends = options.depends ?? {};
  const inputs = options.inputs ?? {};
  const version = options.version ?? '1.0.0';

  const HandlerClass = class extends StepHandler {
    static override handlerName = name;
    static override handlerVersion = version;

    async call(context: StepContext): Promise<StepHandlerResult> {
      try {
        const args = injectArgs(context, depends, inputs);
        const rawResult = await fn(args);
        return wrapResult(rawResult);
      } catch (error: unknown) {
        return wrapError(error);
      }
    }
  };

  Object.defineProperty(HandlerClass, 'name', { value: name });
  return HandlerClass as typeof StepHandler & { new (): StepHandler };
}

/**
 * Define a decision handler from a function.
 *
 * The function should return a `Decision.route(...)` or `Decision.skip(...)`.
 *
 * @param name - Handler name
 * @param options - Dependencies, inputs, and version
 * @param fn - Handler function returning a Decision
 * @returns StepHandler subclass
 *
 * @example
 * ```typescript
 * const RouteOrder = defineDecisionHandler('route_order', {
 *   depends: { order: 'validate_order' },
 * }, async ({ order }) => {
 *   if (order.tier === 'premium') {
 *     return Decision.route(['process_premium'], { tier: 'premium' });
 *   }
 *   return Decision.route(['process_standard']);
 * });
 * ```
 */
export function defineDecisionHandler(
  name: string,
  options: HandlerOptions,
  fn: (args: HandlerArgs) => Promise<Decision | StepHandlerResult>
): typeof StepHandler & { new (): StepHandler } {
  const depends = options.depends ?? {};
  const inputs = options.inputs ?? {};
  const version = options.version ?? '1.0.0';

  const HandlerClass = class extends StepHandler {
    static override handlerName = name;
    static override handlerVersion = version;

    get capabilities(): string[] {
      return ['process', 'decision', 'routing'];
    }

    async call(context: StepContext): Promise<StepHandlerResult> {
      try {
        const args = injectArgs(context, depends, inputs);
        const rawResult = await fn(args);

        if (rawResult instanceof StepHandlerResult) {
          return rawResult;
        }

        if (rawResult instanceof Decision) {
          // Delegate to DecisionMixin for correct serialization format
          const mixin = new DecisionMixin();
          // Bind name/version for metadata generation
          Object.defineProperty(mixin, 'name', { get: () => name });
          Object.defineProperty(mixin, 'version', { get: () => version });

          if (rawResult.type === 'create_steps') {
            return mixin.decisionSuccess(rawResult.steps, rawResult.routingContext);
          }
          return mixin.skipBranches(rawResult.reason ?? '', rawResult.routingContext);
        }

        return wrapResult(rawResult);
      } catch (error: unknown) {
        return wrapError(error);
      }
    }
  };

  Object.defineProperty(HandlerClass, 'name', { value: name });
  return HandlerClass as typeof StepHandler & { new (): StepHandler };
}

/**
 * Define a batch analyzer handler.
 *
 * The function should return a `BatchConfig` with `totalItems` and `batchSize`.
 * Cursor configs are generated automatically.
 *
 * @param name - Handler name
 * @param options - Handler options plus `workerTemplate`
 * @param fn - Handler function returning a BatchConfig
 * @returns StepHandler subclass
 */
export function defineBatchAnalyzer(
  name: string,
  options: HandlerOptions & { workerTemplate: string },
  fn: (args: HandlerArgs) => Promise<BatchConfig | StepHandlerResult>
): typeof StepHandler & { new (): StepHandler } {
  const depends = options.depends ?? {};
  const inputs = options.inputs ?? {};
  const version = options.version ?? '1.0.0';
  const workerTemplate = options.workerTemplate;

  const HandlerClass = class extends StepHandler {
    static override handlerName = name;
    static override handlerVersion = version;

    async call(context: StepContext): Promise<StepHandlerResult> {
      try {
        const args = injectArgs(context, depends, inputs);
        const rawResult = await fn(args);

        if (rawResult instanceof StepHandlerResult) {
          return rawResult;
        }

        const config = rawResult as BatchConfig;
        const { totalItems, batchSize } = config;

        // Delegate cursor generation to BatchableMixin
        const batchMixin = new BatchableMixin();
        const cursorRanges = batchMixin.createCursorRanges(totalItems, batchSize);

        // Convert CursorConfig[] to RustCursorConfig[] for FFI boundary
        const rustConfigs: RustCursorConfig[] = cursorRanges.map((c, idx) => ({
          batch_id: `batch_${String(idx).padStart(3, '0')}`,
          start_cursor: c.startCursor,
          end_cursor: c.endCursor,
          batch_size: c.endCursor - c.startCursor,
        }));

        // Use typed FFI outcome factory
        const batchProcessingOutcome = createBatchesOutcome(
          workerTemplate,
          rustConfigs.length,
          rustConfigs,
          totalItems
        );

        return StepHandlerResult.success(
          { batch_processing_outcome: batchProcessingOutcome },
          {
            batch_count: rustConfigs.length,
            total_items: totalItems,
            ...(config.metadata ?? {}),
          }
        );
      } catch (error: unknown) {
        return wrapError(error);
      }
    }
  };

  Object.defineProperty(HandlerClass, 'name', { value: name });
  return HandlerClass as typeof StepHandler & { new (): StepHandler };
}

/**
 * Define a batch worker handler.
 *
 * The function receives a `batchContext` parameter extracted from the step
 * context, containing cursor configuration for this worker's partition.
 *
 * @param name - Handler name
 * @param options - Dependencies, inputs, and version
 * @param fn - Handler function receiving batch context
 * @returns StepHandler subclass
 */
export function defineBatchWorker(
  name: string,
  options: HandlerOptions,
  fn: (
    args: HandlerArgs & { batchContext: BatchWorkerContext | null }
  ) => Promise<Record<string, unknown> | StepHandlerResult | undefined>
): typeof StepHandler & { new (): StepHandler } {
  const depends = options.depends ?? {};
  const inputs = options.inputs ?? {};
  const version = options.version ?? '1.0.0';

  const HandlerClass = class extends StepHandler {
    static override handlerName = name;
    static override handlerVersion = version;

    async call(context: StepContext): Promise<StepHandlerResult> {
      try {
        const args = injectArgs(context, depends, inputs) as HandlerArgs & {
          batchContext: BatchWorkerContext | null;
        };

        // Delegate batch context extraction to BatchableMixin
        const batchMixin = new BatchableMixin();
        args.batchContext = batchMixin.getBatchContext(context);

        const rawResult = await fn(args);
        return wrapResult(rawResult);
      } catch (error: unknown) {
        return wrapError(error);
      }
    }
  };

  Object.defineProperty(HandlerClass, 'name', { value: name });
  return HandlerClass as typeof StepHandler & { new (): StepHandler };
}

// ============================================================================
// API Handler Options
// ============================================================================

/**
 * Options for defineApiHandler.
 */
export interface ApiHandlerOptions extends HandlerOptions {
  /** Base URL for API calls */
  baseUrl: string;
  /** Default request timeout in milliseconds (default: 30000) */
  defaultTimeout?: number;
  /** Default headers to include in all requests */
  defaultHeaders?: Record<string, string>;
}

/**
 * The injected arguments for API handlers, including the `api` object
 * with pre-configured HTTP methods and result helpers.
 *
 * `api` is the handler instance itself with API methods applied (matching
 * the Python/Ruby pattern where `api=self`).
 */
export type ApiHandlerArgs = HandlerArgs & {
  api: APICapable;
};

// ============================================================================
// API Handler Factory
// ============================================================================

/**
 * Define an API handler with HTTP client functionality.
 *
 * The function receives an `api` object providing pre-configured HTTP methods
 * (get, post, put, patch, delete, request) and result helpers (apiSuccess,
 * apiFailure, connectionError, timeoutError) from the APIMixin.
 *
 * @param name - Handler name (must match step definition)
 * @param options - Dependencies, inputs, version, and API configuration
 * @param fn - Handler function receiving api and other injected args
 * @returns StepHandler subclass
 *
 * @example
 * ```typescript
 * const FetchUser = defineApiHandler('fetch_user', {
 *   baseUrl: 'https://api.example.com',
 *   depends: { userId: 'validate_user' },
 * }, async ({ userId, api }) => {
 *   const response = await api.get(`/users/${userId}`);
 *   if (response.ok) {
 *     return api.apiSuccess(response);
 *   }
 *   return api.apiFailure(response);
 * });
 * ```
 */
export function defineApiHandler(
  name: string,
  options: ApiHandlerOptions,
  fn: (args: ApiHandlerArgs) => Promise<Record<string, unknown> | StepHandlerResult | undefined>
): typeof StepHandler & { new (): StepHandler } {
  const depends = options.depends ?? {};
  const inputs = options.inputs ?? {};
  const version = options.version ?? '1.0.0';
  const apiBaseUrl = options.baseUrl;
  const apiTimeout = options.defaultTimeout ?? 30000;
  const apiHeaders = options.defaultHeaders ?? {};

  const HandlerClass = class extends StepHandler {
    static override handlerName = name;
    static override handlerVersion = version;
    static baseUrl = apiBaseUrl;
    static defaultTimeout = apiTimeout;
    static defaultHeaders = apiHeaders;

    constructor() {
      super();
      // Apply HTTP methods (get, post, put, patch, delete, request) and
      // result helpers (apiSuccess, apiFailure, connectionError, timeoutError)
      // directly onto this instance — matching the Python/Ruby pattern
      // where api=self.
      applyAPI(this);
    }

    async call(context: StepContext): Promise<StepHandlerResult> {
      try {
        const args = injectArgs(context, depends, inputs) as ApiHandlerArgs;
        args.api = this as unknown as APICapable;
        const rawResult = await fn(args);
        return wrapResult(rawResult);
      } catch (error: unknown) {
        return wrapError(error);
      }
    }
  };

  Object.defineProperty(HandlerClass, 'name', { value: name });
  return HandlerClass as typeof StepHandler & { new (): StepHandler };
}
